phosphor-auditlog/auditLogParser.cpp

```diff
@@ -0,0 +1,422 @@
+
+    switch (recType)
+    {
+        case AUDIT_USYS_CONFIG:
+            fillUsysEntry(parsedEntry);
+            break;
+
+        default:
+            fillAuditEntry(parsedEntry);
+            break;
+    }
+
+#ifdef AUDITLOG_FULL_DEBUG
+    lg2::debug("parsedEntry = {PARSEDENTRY}", "PARSEDENTRY",
+               parsedEntry.dump());
+#endif // AUDITLOG_FULL_DEBUG
+
+    return true;
+}
+
+bool ALParser::formatGen(nlohmann::json& parsedEntry)
+{
+    auto fullTimestamp = auparse_get_timestamp(au);
+    if (fullTimestamp == nullptr)
+    {
+        lg2::error("Failed to parse timestamp");
+        throw sdbusplus::xyz::openbmc_project::Common::Error::InternalFailure();
+        return false;
+    }
+    parsedEntry["EventTimestamp"] = fullTimestamp->sec;
+    parsedEntry["ID"] = std::format("{}.{}:{}", fullTimestamp->sec,
+                                    fullTimestamp->milli,
+                                    fullTimestamp->serial);
+
+    auto recMsg = auparse_get_record_text(au);
+    parsedEntry["Event"] = recMsg;
+
+#ifdef AUDITLOG_FULL_DEBUG
+    lg2::debug("parsedEntry = {PARSEDENTRY}", "PARSEDENTRY",
+               parsedEntry.dump());
+#endif // AUDITLOG_FULL_DEBUG
+
+    return true;
+}
+
+bool ALParser::formatRaw(nlohmann::json& parsedEntry)
+{
+    auto recMsg = auparse_get_record_text(au);
+    parsedEntry["Event"] = recMsg;
+
+#ifdef AUDITLOG_FULL_DEBUG
+    lg2::debug("parsedEntry = {PARSEDENTRY}", "PARSEDENTRY",
+               parsedEntry.dump());
+#endif // AUDITLOG_FULL_DEBUG
+
+    return true;
+}
+
+void ALParser::parseRecord()
+{
+    nlohmann::json parsedEntry;
+
+    if (formatEntry(parsedEntry))
+    {
+        // Dump JSON object to parsedStream
+        parsedStream << parsedEntry.dump() << '\n';
+    }
+
+    return;
+}
+
+bool ALParser::openParsedFile(const std::string& filePath)
+{
+    std::error_code ec;
+
+    /* Expect the file has already been created */
+    if (!std::filesystem::exists(filePath, ec))
+    {
+        lg2::error("File {FILE} doesn't already exist.", "FILE", filePath);
+        return false;
+    }
+
+    // Create/Open file using truncate
+    parsedStream.open(filePath, std::ios::out);
+    if (parsedStream.fail())
+    {
+        lg2::error("Failed to open {FILE}", "FILE", filePath);
+        throw sdbusplus::xyz::openbmc_project::Common::File::Error::Open();
+        return false;
+    }
+
+    return true;
+}
+
+void ALParser::processEvents()
+{
+    // Loop over all the events
+    while (auditNextEvent())
+    {
+        parseEvent();
+    }
+}
+
+void ALParser::doParse()
+{
+    lg2::debug("Parsing All");
+    processEvents();
+}
+
+void ALParseLatest::parseRecord()
+{
+    nlohmann::json parsedEntry;
+
+    if (formatEntry(parsedEntry))
+    {
+        // Keep list limited to maxLeftCount
+        if (parsedEntries.size() >= maxLeftCount)
+        {
+            parsedEntries.pop_back();
+        }
+
+        parsedEntries.push_front(parsedEntry.dump());
```

Comment:
Can you share why emplace_front() would be a better choice here than push_front()? Reading the man pages for them I don't see the benefit.

Reply:
push_front() is going to use the copy-semantics from the argument to the object.
emplace_front() can optimize by using move-semantic if possible.

For example, `parsedEntry.dump()` will be moved to the parsedEntries vector instead of creating another temp string object if we use emplace_front().

Reply:
Thank you, I see the benefit of using the move-semantics. I will make this change.